node/src/store/in_memory_store.rs

```diff
@@ -34,65 +42,70 @@ impl InMemoryStore {
             headers: DashMap::new(),
             sampling_data: DashMap::new(),
             height_to_hash: DashMap::new(),
-            head_height: AtomicU64::new(0),
+            stored_ranges: RwLock::new(HeaderRanges::default()),
+            lowest_unsampled_height: AtomicU64::new(1),
             header_added_notifier: Notify::new(),
         }
     }
 
     #[inline]
-    fn get_head_height(&self) -> Result<u64> {
-        let height = self.head_height.load(Ordering::Acquire);
-
-        if height == 0 {
-            Err(StoreError::NotFound)
-        } else {
-            Ok(height)
-        }
+    async fn get_head_height(&self) -> Result<u64> {
+        self.stored_ranges
+            .read()
+            .await
+            .head()
+            .ok_or(StoreError::NotFound)
     }
 
-    pub(crate) fn append_single_unchecked(&self, header: ExtendedHeader) -> Result<()> {
-        let hash = header.hash();
-        let height = header.height().value();
-        let head_height = self.get_head_height().unwrap_or(0);
+    pub(crate) async fn insert(
+        &self,
+        headers: Vec<ExtendedHeader>,
+        verify_neighbours: bool,
+    ) -> Result<()> {
+        let (Some(head), Some(tail)) = (headers.first(), headers.last()) else {
+            return Ok(());
+        };
 
-        // A light check before checking the whole map
-        if head_height > 0 && height <= head_height {
-            return Err(StoreError::HeightExists(height));
-        }
+        let headers_range = head.height().value()..=tail.height().value();
+        println!("I: {headers_range:?}");
+        let neighbours_exist = self.try_insert_to_range(headers_range).await?;
```

Comment:
Because of this we have a race with `insert` and `contains_height`. To avoid the race, we need to keep the lock until the end, which makes `DashMap` unneeded.

Should we drop `DashMap` overall and have a single `RwLock` for the whole thing?

Reply:
You're right, the lock scope change during refactor.

Switched to having the header part (`headers`, `height_to_hash` and `stored_ranges`) in `InMemoryStoreInner` and behind RwLock. I've left sampling data as is - due to append-only nature of the store I think we're fine relying on just `DashMap` there.